File: coding/lexer/python/lexer.py
import re
import sys


exprs = ['\\d+', '\\w+', '[\\+-=]']
names = ['INT',  'NAME', 'SYMBOL']


def main():
    rules = []
    for i, expr in enumerate(exprs):
        rules.append(re.compile('^' + expr))
        print(names[i], rules[-1].pattern)

    print('-' * 32)
    for row, code in enumerate(sys.argv[1:]):
        position = 0
        while True:
            while position < len(code) and (code[position] == ' ' or code[position] == '\t'):
                position += 1
            if position >= len(code):
                break

            source = ''
            tokenType = -1
            for i, rule in enumerate(rules):
                result = rule.findall(code[position:])
                if len(result) > 0:
                    source = result[0]
                    tokenType = i
                    break
            if tokenType >= 0:
                print(f'{names[tokenType]}\t`{source}`\t{row}\t{position}')
                position += len(source)
            else:
                print(f'error in {row}, {position}')


if __name__ == "__main__":
    main()

File: coding/lexer/go/lexer.go
package main

import (
	"fmt"
	"regexp"
	"unicode/utf8"
	"os"
)

var exprs = []string{"\\d+", "[\\p{L}\\d_]+", "[\\+-=]"}
var names = []string{"INT",  "NAME",         "SYMBOL"}

func main() {
	rules := []*regexp.Regexp{}
	for i, expr := range exprs {
		rule, _ := regexp.Compile("^" + expr)
		rules = append(rules, rule)
		fmt.Println(names[i], rule)
	}

	fmt.Println("--------------------------------")
	for row, code := range os.Args[1:] {
		position := 0
		col := 0
		for true {
			for position < len(code) && (code[position] == ' ' || code[position] == '\t') {
				position += 1
				col += 1
			}
			if position >= len(code) {
				break
			}
			source := ""
			tokenType := -1
			for i, rule := range rules {
				source = rule.FindString(code[position:])
				if source != "" {
					tokenType = i
					break
				}
			}
			if tokenType >= 0 {
				fmt.Printf("%s\t`%s`\t%d\t%d\n", names[tokenType], source, row, col)
				position += len(source)
				col += utf8.RuneCountInString(source)
			} else {
				fmt.Printf("error in: %d, %d\n", row, col)
				break
			}
		}
	}

}
